### PR TITLE
network mod

### PR DESCRIPTION
**Pull request by quininer**
It lacks documentation and test. @_@

### CODE REVIEW

File: src/toxcore/network.rs [hunk 1]
```diff
@@ -40,11 +176,9 @@ pub const PORT_MAX: u16 = 33545;
           - if 2 args in closure, validate port range from args and try to use
             them to do the binding
 */
```
Comment (zetok) [inline src/toxcore/network.rs]: This `TODO` can be now removed :)

**Comment by zetok** [comment]
From what I understand, `question_mark` just provides syntactic sugar, without which there would be (a lot) more code, but it would build on stable Rust, right?

I'd prefer if things worked at least on latest stable Rust, so that people could build it without much of a hassle, and bump minimal required Rust version only when newer version provides stuff that is necessary to make things work. If nightly is required for some stuff that is optional, by default it shouldn't be required.

From what I see, `question_mark` is going to be stabilized in 1.9, the next release. Current options are:
1. make it work on current stable (1.8) by adding more code, and reverting those changes once 1.9 gets stabilized
2. wait with merging this PR until 1.9 gets stabilized
3. bump minimal required version to current rust beta (1.9), which should be able to compile the code

Personally, I'd prefer to go with `1` (that is, assuming that there are no other things in this PR that would require nightly Rust).

Which option do you think is the best?

File: src/toxcore/network.rs [hunk 2]
```diff
@@ -22,13 +22,149 @@
...
+use std::net::{ UdpSocket, SocketAddr, IpAddr, Ipv6Addr, ToSocketAddrs };
+use std::collections::HashMap;
+use sodiumoxide;
 
 
 /// Minimum port which Tox will try to bind to.
 pub const PORT_MIN: u16 = 33445;
 /// Maximum port which Tox will try to bind to.
 pub const PORT_MAX: u16 = 33545;
+pub const MAX_UDP_PACKET_SIZE: usize = 2048;
```
Comment (zetok) [inline src/toxcore/network.rs:42]:
```
/// Maximum size of a UDP packet that tox will handle.
```

?

**Comment by zetok** [comment]
Oh, right, the other thing that won't work on 1.8 / 1.9 are block docs for stuff in `impl` block – that has been fixed (IIRC) only in nightly.

**Comment by quininer** [comment]
I like `?` syntax, because I only use rust-nightly. but I do not mind using a `try!`.
In addition, using `mio` instead of `std::net` is better?

**Comment by zetok** [comment]
> In addition, using `mio` instead of `std::net` is better?

I think so :)

I haven't looked into that much, but generally using a widely used lib that seems to be quite good most likely will be better than re-inventing the wheel.

**Comment by quininer** [comment]
when I have time, and then try to migrate to it `mio`. :)

**Comment by zetok** [comment]
Since functionality requires beta, you'd want to remove this line https://github.com/zetok/tox/blob/b2b934e2bc3dbbaf04bc087a06e2a93d25f967c7/.travis.yml#L3 and add info in `Building` section of README.md that `Rust beta (1.9)` is required.

**Comment by zetok** [comment]
Could you perhaps add documentation in form of the example usage code?

E.g. https://github.com/zetok/tox/blob/85508e1293d7f4dde87511b0cda3fa94fff5138c/src/toxcore/state_format/old.rs#L204,L257

It doesn't need to be that long, but it should at least test whether implementation works.

**Comment by quininer** [comment]
I try to use mio: https://github.com/quininer/tox/tree/mio

**Comment by zetok** [comment]
It can be fast-forwarded into this PR, right?

I.e. it does seem to work, so..

**Comment by zetok** [comment]
- [x] since `MAX_UDP_PACKET_SIZE` is now defined, it should be used in https://github.com/zetok/tox/pull/8/files#diff-b4aea3e418ccdb71239b96952d9cddb6R85

**Comment by zetok** [comment]
- [ ] all warnings need to be fixed
  
  Since they're about missing docs, they should be written and/or fixed. There are quite a few things where docs are outright wrong, and that's bad.
  
  Generally, docs alone should be enough to make it possible to use API – currently one would have to read the code – that's the lowest bar that needs to be met before PR gets merged.

You've said that your English isn't that good, and you'd want help with writing docs – I can help with expanding existing docs and writing some for the stuff that is outright clear, but other stuff requires you to write docs.

Even if the docs that you'll write won't be good/clear on the first try, they can always be fixed and improved, and any docs are better than no docs at all :)

Also, note that I'm not particularly good at English myself, so there will be some things that I'll miss :)

So, it would be good if you could add as much docs as you can to the code (even if the way they'd be written wouldn't be of high quality), and if that won't cover all the code, just make comments on things explaining how they are supposed to work/for what they're needed, and I'll try to make docs out of it.

If something won't be clear, I'll just ask :)

**Comment by zetok** [comment]
Hmm, since now 1.9 is stable, and beta is failing, 4f0851e570f098d67224842b5e11d7020b70cf10 should be removed.

File: src/lib.rs [hunk 3]
```diff
@@ -62,7 +62,7 @@ fn main() {
 
     // and since packet is ready, prepare the network part;
     // bind to some UDP socket
```
Comment (zetok) [inline src/lib.rs]: Should be (something like that?):
`bind to UDP socket for supplied address and port in range`

File: src/toxcore/network.rs [hunk 4]
```diff
@@ -22,29 +22,205 @@
...
+/** Run this before creating sockets.
+
+    return true on success
+    return false on failure
+*/
+fn networking_at_startup() -> bool {
+    let mut output = true;
+    AT_STARTUP_RAN.call_once(|| output = sodiumoxide::init());
+    output
+}
```
Comment (zetok) [inline src/toxcore/network.rs]: Initializing libsodium shouldn't happen in networking code.

Comment (quininer) [inline src/toxcore/network.rs]: https://github.com/irungentoo/toxcore/blob/master/toxcore/network.c#L439

I also think that it should not be here, but it should be placed where it?

Comment (zetok) [inline src/toxcore/network.rs]: most likely `crypto_core`

**Pull request merged**